### PR TITLE
Refactor Apex and LWC tests

### PR DESCRIPTION
**Pull request by CristiCanizales**
Refactor to reduce dupe code in lwc and apex tests

@W-16688177@

[LWC End to End Tests](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/10765914393) ✅ 
[Apex End to End Tests](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/10765912870) ✅

### CODE REVIEW

File: test/utilities/testsUtils.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  return testsItems;
+}
+
+export async function getTestsSection(workbench: Workbench, type: string) {
+  const sidebar = workbench.getSideBar();
+  const sidebarView = sidebar.getContent();
+  const testsSection = await sidebarView.getSection(type);
+  await expect(testsSection.elem).toBeDefined();
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:48]: Why did you change this from `toBePresent()` to `toBeDefined()`?

File: test/utilities/testsUtils.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+ *   1
+ * );
+ */
+export async function verifyTestItemsInSideBar(
+  testsSection: ViewSection,
+  refreshCommand: string,
+  expectedItems: string[],
+  expectedTests: number,
+  expectedClasses: number
+) {
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:191]:
```suggestion
): Promise<void> {
```

To get rid of the warning from the 3 dots
Screenshot 2024-09-09 at 10 15 29 AM

File: test/utilities/testsUtils.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+ *   2,
+ *   1
+ * );
+ */
+export async function verifyTestItemsInSideBar(
+  testsSection: ViewSection,
+  refreshCommand: string,
+  expectedItems: string[],
+  expectedTests: number,
+  expectedClasses: number
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:190]:
```suggestion
  expectedNumTests: number,
  expectedNumClasses: number
```

To match the variable naming in `retrieveExpectedNumTestsFromSidebar()`, which I found to be more clear.

File: test/utilities/testsUtils.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+) {
+  log('Starting verifyTestItemsInSideBar()');
+  const testsItems = await retrieveExpectedNumTestsFromSidebar(
+    expectedTests,
+    testsSection,
+    refreshCommand
+  );
+  const isLWCSection = refreshCommand.includes('Lightning');
+  if (isLWCSection) {
+    log('Running LWC Tests');
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:200]:
```suggestion
    log('Expanding LWC Tests');
```

File: test/utilities/testsUtils.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    // Expand tests
+    for (let x = 0; x < expectedClasses; x++) {
+      await testsItems[x].expand();
+    }
+  }
+
+  // Make sure all the tests are present in the sidebar
+  await expect(testsItems.length).toBe(isLWCSection ? expectedClasses : expectedTests);
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:208]: Given that the LWC Tests section is expanded in Lines 199-205, this doesn't make sense to me.  Why aren't we checking for `expectedTests` for both the Apex Tests and LWC Tests sections?

File: test/specs/debugLwcTests.e2e.ts [hunk 6]
```diff
@@ -59,60 +58,49 @@ describe('Debug LWC Tests', async () => {
...
     );
-    // Expand LWC tests
-    await lwcTestsItems[0].expand();
-    await lwcTestsItems[1].expand();
 
     // Click the debug test button that is shown to the right when you hover a test class name on the Test sidebar
-    const lwcTestItem = (await lwcTestsSection.findItem('lwc1')) as TreeItem;
-    await lwcTestItem.select();
-    const debugTestsAction = await lwcTestItem.getActionButton(
+    await lwcTestsItems[0].select();
```
Comment (daphne-sfdc) [inline test/specs/debugLwcTests.e2e.ts:71]: Nice!  This is much cleaner than before 😀

File: test/specs/debugLwcTests.e2e.ts [hunk 7]
```diff
@@ -59,60 +58,49 @@ describe('Debug LWC Tests', async () => {
...
+    const debugTestsAction = await lwcTestsItems[0].getActionButton(
       'SFDX: Debug Lightning Web Component Test File'
     );
-    if (!debugTestsAction) {
-      fail('Could not find debug tests action button');
-    }
-    await debugTestsAction.elem.click();
+    await expect(debugTestsAction).toBeDefined();
```
Comment (daphne-sfdc) [inline test/specs/debugLwcTests.e2e.ts:75]: I like this change 🙂

File: test/specs/debugLwcTests.e2e.ts (hunk 6 above)
Comment (daphne-sfdc) [inline test/specs/debugLwcTests.e2e.ts:71]: Nice!  This is much cleaner than before 😀

File: test/specs/debugLwcTests.e2e.ts (hunk 7 above)
Comment (daphne-sfdc) [inline test/specs/debugLwcTests.e2e.ts:75]: Another great refactor 🙂

File: test/utilities/testsUtils.ts (hunk 1 above)
Comment (CristiCanizales) [reply test/utilities/testsUtils.ts:48]: Because of [this](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/10765701306/job/29850338426#step:13:355) error

File: test/specs/debugLwcTests.e2e.ts [hunk 8]
```diff
@@ -59,60 +58,49 @@ describe('Debug LWC Tests', async () => {
...
-    );
+    const expectedTexts = [
+      'PASS force-app/main/default/lwc/lwc1/__tests__/lwc1.test.js',
+      'Test Suites: 1 passed, 1 total',
+      'Tests:       2 passed, 2 total',
+      'Snapshots:   0 total',
+      `Ran all test suites within paths "${path.join(projectFolderPath, 'force-app', 'main', 'default', 'lwc', 'lwc1', '__tests__', 'lwc1.test.js')}"`
+    ];
+    await expect(terminalText).toBeDefined();
+    await utilities.verifyTestResult(terminalText!, expectedTexts);
```
Comment (daphne-sfdc) [inline test/specs/debugLwcTests.e2e.ts:96]: I love the new `verifyTestResult()` helper function! 💯

File: test/utilities/testsUtils.ts (hunk 5 above)
Comment (CristiCanizales) [reply test/utilities/testsUtils.ts:208]: They behave differently🤷🏻‍♀️ When you refresh apex tests, they show up expanded. When you refresh LWC tests, they show up collapsd. It's the same expectation than before the refactor, I didn't change that. That will have to change with [W-16428742](https://gus.lightning.force.com/a07EE00001yKRUPYA4).

File: test/utilities/testsUtils.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+export async function verifyTestIconColor(testItem: TreeItem, colorLabel: string) {
+  log(`Verifying icon's colors - verifyTestIconColor()`);
+  // Verify the tests that are passing are labeled with a green dot on the Test sidebar
+  const icon = await (await testItem.elem).$('.custom-view-tree-node-item-icon');
+  const iconStyle = await icon.getAttribute('style');
+  // Try/catch used to get around arbitrary flaky failure on Ubuntu in remote
+  try {
+    await expect(iconStyle).toContain(colorLabel);
+  } catch {
+    log('ERROR: icon did not turn green after test successfully ran');
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:135]:
```suggestion
    log(`ERROR: icon color is not ${colorLabel}`);
```

Also, I see the above comment that the try/catch was used to get around an arbitrary flaky failure in Ubuntu.  Can we verify whether the arbitrary flaky failure still exists and remove the try/catch if it's no longer an issue?

Comment (CristiCanizales) [reply test/utilities/testsUtils.ts:135]: I think we should keep it. We have not done any changes around that and as it says, it's flaky. Maybe it will pass in a couple of runs and then just start failing again, and we don't want to introduce more flappers :dead1:

File: test/utilities/testsUtils.ts (hunk 5 above)
Comment (daphne-sfdc) [reply test/utilities/testsUtils.ts]: Ah I see.  `testsItems` is defined as the number of entries in the sidebar, which is only the LWC classes because the section is collapsed.  Lines 199-205 to expand each section comes _after_ `testsItems` is defined, and `testsItems` is not modified to include the new entries after expansion. ✅

File: test/specs/runApexTests.e2e.ts [hunk 10]
```diff
@@ -253,136 +263,72 @@ describe('Run Apex Tests', async () => {
...
-    await expect(outputPanelText).toContain('ended SFDX: Run Apex Tests');
+    const expectedTexts = [
+      '=== Test Summary',
+      'Outcome              Passed',
+      'Tests Ran            3',
+      'Pass Rate            100%',
+      'TEST NAME',
+      'ExampleApexClass1Test.validateSayHello  Pass',
+      'ExampleApexClass2Test.validateSayHello  Pass',
+      'ExampleApexClass3Test.validateSayHello  Pass',
```
Comment (daphne-sfdc) [inline test/specs/runApexTests.e2e.ts:274]: Good catch! 😁

File: test/utilities/testsUtils.ts [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+ * @example
+ * const result = await runTestCaseFromSideBar(
+ *   myWorkbench,
+ *   'APEX TESTS',
+ *   'MyApexTestCase',
+ *   'SFDX: Run Apex Tests'
+ * );
+ * console.log(result); // Outputs the result from the Apex test run
+ */
+export async function runTestCaseFromSideBar(
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:69]: Love this new helper function! ❤️

File: test/utilities/testsUtils.ts [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  const actionButton = await testItem.getActionButton(actionLabel);
+  await expect(actionButton).toBeDefined();
+  await actionButton?.elem.click();
+
+  if (actionLabel.includes('Debug')) {
+    await pause(Duration.seconds(10));
+
+    // Continue with the debug session
+    await continueDebugging();
+  }
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:99]: Is this block needed?  I see that the `runTestCaseFromSideBar()` helper function currently isn't being used for the DebugApexTests and DebugLWCTests E2E tests.

Comment (CristiCanizales) [reply test/utilities/testsUtils.ts:99]: Oh good catch, I was trying some stuff and forgot to remove this

File: test/utilities/testsUtils.ts [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+  return testsSection;
+}
+
+/**
+ * Runs a test case from the sidebar and returns the test result.
+ * *
+ * @param {Workbench} workbench - The workbench instance used to interact with the sidebar and views.
+ * @param {string} testSuite - The name of the test suite from which to run the test (e.g., 'APEX TESTS', 'LWC TESTS').
+ * @param {string} testName - The name of the specific test case to run.
+ * @param {string} actionLabel - The label of the action button to click (e.g., 'SFDX: Run Apex Tests', 'Debug').
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:58]:
```suggestion
 * @param {string} actionLabel - The label of the action button to click (e.g., 'Run Single Test', 'SFDX: Run Lightning Web Component Test File').
```

This function is used only for run, not debug.  Also, the buttons for Apex are "Run Tests" and "Run Single Test".

File: test/utilities/testsUtils.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+ *
+ * @example
+ * const result = await runTestCaseFromSideBar(
+ *   myWorkbench,
+ *   'APEX TESTS',
+ *   'MyApexTestCase',
+ *   'SFDX: Run Apex Tests'
```
Comment (daphne-sfdc) [inline test/utilities/testsUtils.ts:65]:
```suggestion
 *   'Run Single Test'
```

File: test/specs/runLwcTests.e2e.ts [hunk 15]
```diff
@@ -273,63 +176,28 @@ describe('Run LWC Tests', async () => {
...
-
-    // Verify the tests that are passing are labeled with a green dot on the Test sidebar
-    const icon = await (await lwcTestItem.elem).$('.custom-view-tree-node-item-icon');
-    const iconStyle = await icon.getAttribute('style');
-    // Try/catch used to get around arbitrary flaky failure on Ubuntu in remote
-    try {
-      await expect(iconStyle).toContain('testPass');
-    } catch {
-      utilities.log('ERROR: icon did not turn green after test successfully ran');
-    }
```
[the remark this answers is not in the document]
Comment (CristiCanizales) [reply test/specs/runLwcTests.e2e.ts:287]: because this is covered in runTestCaseFromSideBar

File: test/specs/runLwcTests.e2e.ts [hunk 16]
```diff
@@ -339,21 +207,15 @@ describe('Run LWC Tests', async () => {
...
```
[the remark this answers is not in the document]
Comment (CristiCanizales) [reply test/specs/runLwcTests.e2e.ts:353]: same, this is covered in runTestCaseFromSideBar

**Review by daphne-sfdc — APPROVED** [review]
Approved!  Thank you @CristiCanizales for the great refactoring! 🎉 🥳 

Passing E2E test runs:
- Apex: https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/10780713719
- LWC: https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/10780742454
